### PR TITLE
Use setuptools <74 and buildout 3.1

### PR DESCRIPTION
**Pull request by dataflake**

### CODE REVIEW

**Review by icemac — DISMISSED** [review]

File: config/c-code/tests.yml.j2 [hunk 1]
```diff
@@ -104,15 +104,15 @@ jobs:
   {% if require_cffi %}
           pip install -U --pre cffi
   {% endif %}
-          pip install -U "setuptools<69" wheel twine
+          pip install -U "setuptools<74" wheel twine
```
[the remark this answers is not in the document]
Comment (dataflake) [reply config/c-code/tests.yml.j2:107]: You're right, the title is all wrong. But the moment "when it creaks again" is already here with the `cffi` incompatibility.

**Comment by dataflake** [comment]
It seems that there's an ongoing compatibility problem between `zc.buildout` and `setuptools`. Any new setuptools release risks breaking zc.buildout, and as we see repairs take long. We keep introducing and removing setuptools pins as a workaround. I see two different ways to approach the situation:

- every time the latest setuptools works we remove the pin. When it breaks we add one again.
- we always pin setuptools to a version that works at that moment and just move the pin, we never remove it

File: config/c-code/tests.yml.j2 (hunk 1 above)
[the remark this answers is not in the document]
Comment (dataflake) [reply config/c-code/tests.yml.j2:107]: Just to make sure you didn't object to the pin just for the one place you marked, which is the install for future python, I tested that explicitly and it fails (https://github.com/zopefoundation/persistent/actions/runs/10628933215). So the pin must stay here.

**Comment by mauritsvanrees** [comment]
> * we always pin setuptools to a version that works at that moment and just move the pin, we never remove it

This makes sense to me.

The danger is that we keep testing against an older setuptools version for a long time, and are not aware of breakage with newer versions.  But I guess we will see that soon enough in client projects or when someone manually tries to update the pin in one package and it fails.  So should be fine.

**Review by mauritsvanrees — APPROVED** [review]

**Comment by dataflake** [comment]
I agree that always keeping a pin in place is a better solution. I believe it will lead to less unwelcome surprises whenever setuptools has a new release.

**Comment by dataflake** [comment]
I expanded this PR to store the setuptools version specification in one central place and re-use it in the templates. That way no one has to grep the templates whenever the pin changes.

**Comment by tseaver** [comment]
Could ship with one testing environment that runs without the pin as a "canary in the coal mine", so that we know when things break, and flag that environment as OK to fail?   Maybe latest released Python, only on Ubuntu?

**Comment by dataflake** [comment]
@tseaver Take a look at #275 for one way to do this.

**Comment by tseaver** [comment]
@dataflake #275 looks reasonable, but I was hoping it would run under CI, with and "allowed to fail" mode.

**Comment by dataflake** [comment]
I'm just concerned that we're adding more and more runs there. Full runs take very long as is.

**Comment by tseaver** [comment]
@dataflake 
> I'm just concerned that we're adding more and more runs there. Full runs take very long as is.

I get that.  Perhaps we might setup something which runs nightly (using a [schedule event](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#schedule), rather than on a push, and create an issue if there is a failure?  For the last, something like [this example](https://docs.github.com/en/actions/use-cases-and-examples/project-management/scheduling-issue-creation), but only if the tests fail.

**Comment by dataflake** [comment]
Do you want to try your hand at such a workflow?

**Comment by dataflake** [comment]
I have decided to merge this specific part now, let's continue the discussion about testing with the latest setuptools in #275

File: config/c-code/tests.yml.j2 (hunk 1 above)
[the remark this answers is not in the document]
Comment (icemac) [reply config/c-code/tests.yml.j2]: It was mostly for all places but with the `cffi` incompatibility we already need an upper version bound again.

**Pull request merged**